### ISSUE TITLE
Sound split: closing NVDA doesn't restore panning for closed applications

### ISSUE THREAD

**Issue by LeonarddeR** [nvaccess/nvda#16292]
### Steps to reproduce:
1. Start Foobar2000 and play some audio
2. Set sound split in NVDA, e.g. to NVDA left, apps right
3. Close Foobar2000
4. Disable sound split
5. Close NVDA
6. Start Foobar2000 and start playing again.

### Actual behavior:
Foobar is still panned to the right.

### Expected behavior:
Foobar should play through both speakers.

### Additional context
Note that NVDA seems to restore Foobars panning as soon as you start it. This basically means that sound split, even when it is disabled, changes the panning for applications.

#### NVDA installed/portable/running from source:
Installed

#### NVDA version:
alpha-31398

#### Windows version:
Windows 11 23H2

#### Name and version of other software in use when reproducing the issue:
Foobar2000 2.1.2

#### Does the issue still occur after restarting your computer?
Not tested

#### Have you tried any other versions of NVDA? If so, please report their behaviors.
No

#### If NVDA add-ons are disabled, is your problem still occurring?
Yes

#### Does the issue still occur after you run the COM Registration Fixing Tool in NVDA's tools menu?
N/a

**Comment by paulber19** [comment]
Note that NVDA seems to restore Foobars panning as soon as you start it. 
This basically means that sound split, even when it is disabled, changes 
the panning for applications
Thank you Leonard for this remark. This is exactly what I tried to 
convey in Issue #16287
It is not possible to deactivate the sound split functionality.
In this mode and even after starting, NVDA will systematically restore 
the sound of any new audio source to both channels.

**Comment by CyrilleB79** [comment]
Anybody knows how does Jaws behave with its Sound Split feature in this situation?

**Comment by nidza07** [comment]
Hello,
testing the same STR with Jaws, latest version (2024 March update), this issue does not happen.

Enabling sound split, quitting an app, disabling sound split in Jaws, quitting Jaws and then reopening the previously quit app, that app plays in the center.

**Comment by mltony** [comment]
That's an interesting issue. I don't know off the top of my head how to solve it. I'll see if I can subscribe to notifications telling me when AudioSession is about to be terminated and restore its volume.
@nidza07,
Do you know if in jaws there's a readable script that I can take a look at to see how they handle this?

**Comment by LeonarddeR** [comment]
Isn't there a way to reset the audio session state after it was closed?
Still, I think that applications that could possible mutate their own audio sessions should be respected, therefore sound split of should really be off and leave all audio sessions alone.

**Comment by mltony** [comment]
Ok, here is what I found so far:
1. Catching the moment just before the application quits doesn't seem to be supported by wasapi. It can only let you know when audio session has already been terminated, at which moment it is too late to adjust its volume.
2. Setting volume using wasapi is persistent - e.g. application volume is preserved across app restarts and computer restarts. There appears to beno way to make it non-persistent.
3. Wasapi must be storing its internal state somewhere including app volumes. However this doesn't appear to be exposed through any public API, so we can't change it.
4. I digged a bit into Jaws scripts. I found a script called `Script RouteJAWSSoundRight`, which seems to do the job, however it appears to only be setting some global variables. These variables don't appear to be used anywhere else in scripts, which makes me think that they are read by compiled code that we don't have access to.
5. Simple grepping Jaws installation and scripts directory for keywords like IChannelAudioVolume does not return any results. However, grepping for more common things like IAccessible2  doesn't return any results in the binaries either, despite main executable being 20mb. Which makes me think it must be obfuscated somehow. So we can't tell for sure whether Jaws is using wasapi for its sound split or some other technology.
6. IN readable jaws scripts I see some lines like:
    ```
    code == SoundMixerFeedback_RoutedLeft
    ```
    Which kind of implies there is some thing called sound mixer inside jaws. It's hard to make any conclusions based on this circumstantial evidence, but it's possible that Jaws does create a new sound device and it routes all applications through it in order to do sound split. IIRC Jaws does require to restart computer upon installation, which might be another indicator of installing new device drivers. If that's the case, then I doubt that we'd want to take that route, as building new device driver seems to me a huge overhead hardly justified.
    
TL;DR: I couldn't figure out how Jaws does sound split in such a way that it doesn't cause this issue. I don't see a way to fix it using wasapi either. I suppose we can document this in user docs that this is just another quirk of sound split.
Any thoughts?

**Comment by Adriani90** [comment]
cc: @jcsteh maybe you have some suggestions here for https://github.com/nvaccess/nvda/issues/16292#issuecomment-1992522907

**Comment by jcsteh** [comment]
To clarify, after IAudioSessionEvents::OnSessionDisconnected, do you get errors if you try to use IChannelAudioVolume::SetChannelVolume?

**Comment by mltony** [comment]
I did some more investigation. For whatever reason, OnSessionDisconnected is never called at all. However, I figured out that on_state_changed is being called with state= Expired. And yes I can set channel  volume in that call.
Working on PR now, looks like I'll be able to fix this in the end.

**Comment by LeonarddeR** [comment]
Lovely @mltony. May be you could make disabled really disabled in that pr? I.e. disabled should never touch any audio session whatsoever rather than enforcing both channels same volume

**Comment by paulber19** [comment]
I agree. see issue #16287 which was closed for the wrong reason
